PICARD-2339: Faster clustering implementation

phw opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary

<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
  * [x] Bug fix
  * [ ] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-2339
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->

The existing clustering code is using the Levenshtein distance to calculate similarity, which caused a O(n^2) performance. But since only exactly similar matches are being used (similarity threshold 1.0) this was is not strictly necessary.

If we don't use a threshold (and Picard never did) then we can simplify the code.


# Solution

This new implementation uses simple comparison for string equality and performs in O(n).

Added tests to verify consistent behavior.

There is a small change in behavior: Previously if there were multiple different spelling for what is considered the same album title the last encountered spelling would be used. The new implementation uses the first one.

See https://github.com/metabrainz/picard/commit/760af4dc7a45801df2a86eee0afc794d172d26a9#diff-2d1bf33b26a905bd43b846bc0f989f2b3ff96f82da68719d465a330ff0e20ebeL94
<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->
This patch is based on my older work on https://github.com/phw/picard/commit/e31534dbe712b51d0b631e031351be9481a093fa

# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->
We should do a performance comparison.

zas commented on picard/cluster.py:
```diff
@@ -443,183 +415,35 @@ def lookup_metadata(self):
+    def add(self, album, artist, file):
+        self.files.append(file)
+        self._artist_counts[tokenize(artist)] += 1
+        self._artists[tokenize(artist)][artist] += 1
```
I don't think we need `self._artist_counts` since we can use `len(self._artists[artist_token])` instead.

phw replied:
It would actually be `sum(self._artists[artist_token])`. Not sure. We would first need to take the artist with highest sum, and then from that the individual highest "spelling variation". The separate artist_counts looks more efficient. But I'm not really happy about the implementation of the `artist` property. There must be a better way :D

zas replied:
As far I understand this code, `self._artist_counts[tokenize(artist)] == len(self._artists[tokenize(artist)])` but that's not what you want it seems.
If you want artist full name having most occurrences in `self._artists[tokenize(artist)]`, that's `max(self._artists[tokenize(artist)])`.

phw replied:
Still no :) It's a two step process of first choosing the most common artist, and then the most common spelling. We need an example. We have a cluster with different files and the following artist entries:

- artist 1
- ARTIST 1
- ARTIST 1
- artist1
- artist 2
- artist 2
- artist 2

Because of `tokenize` we have two entries in `_artists`: `artist1`, `artist2`.

`artist1` has 4 occurrences, `artist2` has 3. So we want to use `artist1`. That`s currently memorized in `_artist_count`.

`artist1` has 3 different spellings, "artist 1", "ARTIST 1" and "artist1". "ARTIST 1" is the most common, with 2 occurrences. That's what we want to use. These counts of different spellings for `artist1` are currently memorized in `self._artists["artist1"]`.

So to clarify: In the example `self._artist_counts[tokenize("artist1")] == 4`, but `len(self._artists[tokenize("artist1")]) == 3`. 

> If you want artist full name having most occurrences in self._artists[tokenize(artist)], that's max(self._artists[tokenize(artist)], key=self._artists[tokenize(artist)].get).

That's actually nicer then the itemgetter approach I used, so that means it could be:

```
@property
def artist(self):
    tokenized_artist = max(self._artist_counts, key=self._artist_counts.get)
    return max(self._artists[tokenized_artist], key=self._artists[tokenized_artist].get)

@property
def title(self):
    # Find the most common title
    return max(self._titles, key=self._titles.get)
```

zas replied:
Yes, this is what I figured, see irc PM

phw commented:
@zas: I think this is ready know for full review. I extended the tests a bit and added missing behavior for empty artists. The old implementation then hard coded "Various Artists". I made this use the va_name from settings, since this is a slight change in behavior I added PICARD-2340 for this.

Next step would be to run the clustering itself in a thread to offload it from main thread. Although it is now reasonable fast I think we still should do that. It is pure calculation, and the more files you cluster the longer it takes. But I'll do this in a separate PR, not sure what side effects this might have.

zas reviewed (approved):
LGTM, good work!

phw commented on picard/cluster.py:
```diff
@@ -320,61 +324,31 @@ def cluster(files, threshold, tagger=None):
             if win_compat:
                 filename = ntpath.splitdrive(file.filename)[1]
             else:
                 filename = file.filename
```
@zas This is a part I did not touch here, but I wonder about the need for this. If we are on Windows I think `ntpath.splitdrive` should always be done by `album_artist_from_path`, so you don't end up with an artist named "C:".

But do you see any need to have this for other platforms if `config.setting["windows_compatibility"]`  is True? It's really only needed to split this Windows drive letter like "C:\\" or that UNC path to a share, such as "\\\\MyNAS\\theshare\\". But I would not have such paths on e.g. Linux or macOS, and ntpath.splitdrive has no effect on paths like e.g. "/mount/mynas/theshare/xyz".

So I would propose we move the call to `ntpath.splitdrive` to  `album_artist_from_path`, but only check for IS_WIN.

zas replied:
Yes, I agree.
But I would do that in another PR.
